Panic in ValidateAttribute when NEXT_HOP attribute length < 4

A malformed BGP UPDATE message containing a NEXT_HOP path attribute with an invalid length (< 4) can cause GoBGP to panic during attribute validation.

The issue occurs after the UPDATE message is partially accepted and processed, leading to misaligned path attribute parsing and an out-of-bounds access during attribute validation.


**Reproduce Steps**
***
1. Establish an EBGP session with GoBGP.("State":"BGP_FSM_ESTABLISHED",)
2. Send a BGP UPDATE message with:
   - Path Attributes:
     * NEXT_HOP with AttrLen = 0, 1, 2, or 3
     * Payload included
3. gobgpd panics during UPDATE validation.

**Observed Result**
***
```
{"time":"2026-01-27T09:53:58.584154743+08:00","level":"WARN","msg":"the received Update message was treated as withdraw","Topic":"Peer","Key":"192.168.31.195","State":"BGP_FSM_ESTABLISHED","Error":"well-known attribute BGP_ATTR_TYPE_NEXT_HOP must have transitive flag 1"}
panic: runtime error: index out of range [0] with length 0

goroutine 37 [running]:
github.com/osrg/gobgp/v4/pkg/packet/bgp.ValidateAttribute.func2(...)
        /home/runner/work/gobgp/gobgp/pkg/packet/bgp/validate.go:138
github.com/osrg/gobgp/v4/pkg/packet/bgp.ValidateAttribute({0x10ed238, 0xc00035e220}, 0xc000105a0f?, 0x5f?, 0x45?, 0x45?)
        /home/runner/work/gobgp/gobgp/pkg/packet/bgp/validate.go:152 +0x7ae
github.com/osrg/gobgp/v4/pkg/packet/bgp.ValidateUpdateMsg(0xc00022d7a0, 0xc0002fe210, 0x1, 0x0, 0x0)
        /home/runner/work/gobgp/gobgp/pkg/packet/bgp/validate.go:35 +0x3ad
github.com/osrg/gobgp/v4/pkg/server.(*fsmHandler).recvMessageloop(0xc000243080, {0x10e47f0, 0xc0002f6820}, {0x10eea98, 0xc0002a0780}, 0xc000096620, 0xc0002f9b90, 0xc000225dc0?)
        /home/runner/work/gobgp/gobgp/pkg/server/fsm.go:1706 +0x9b7
created by github.com/osrg/gobgp/v4/pkg/server.(*fsmHandler).established in goroutine 32
        /home/runner/work/gobgp/gobgp/pkg/server/fsm.go:1805 +0x2ca
```
**Expected Result**
***
A malformed BGP UPDATE message containing a NEXT_HOP attribute with an invalid length should be safely rejected or cause the affected route to be withdrawn, while the BGP process continues normal operation without crashing.

> According to RFC 4271, the NEXT_HOP attribute is a well-known mandatory path attribute for IPv4 unicast routes and MUST have a length of exactly 4 bytes. A NEXT_HOP attribute with a different length is malformed.
According to RFC 7606, the UPDATE message MUST be handled using the “treat-as-withdraw” approach or be ignored, while keeping the BGP session in the ESTABLISHED state.Malformed path attributes in UPDATE messages should not result in session termination.



**Analysis**
***
GoBGP allows a PathAttribute NextHop with invalid AttrLen (<4) to reach validation, while still constructing a PathAttributeNextHop value, resulting in an internally inconsistent UPDATE state that later causes a panic during AS_PATH validation.

**Suggested Fix**
***
Add length validation for NEXT_HOP attributes during parsing or validation.


**Config:**
```
$ cat gobgp-server1.conf
[global.config]
  as = 65001
  router-id = "192.168.31.142"

[[neighbors]]
  [neighbors.config]
    neighbor-address = "192.168.31.195"
    peer-as = 65002
    [neighbors.transport.config]
    passive-mode = true

```


```
$ ./gobgpd --version
gobgpd version 4.2.0
```


```
$ cat /etc/os-release
PRETTY_NAME="Ubuntu 22.04.5 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.5 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy
```

--- a/pkg/packet/bgp/bgp.go
+++ b/pkg/packet/bgp/bgp.go
@@ -11255,7 +11255,13 @@ func (p *PathAttributeNextHop) DecodeFromBytes(data []byte, options ...*Marshall
 		eSubCode := uint8(BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR)
 		return NewMessageError(eCode, eSubCode, nil, "nexthop length isn't correct")
 	}
-	p.Value, _ = netip.AddrFromSlice(value)
+	var ok bool
+	p.Value, ok = netip.AddrFromSlice(value)
+	if !ok {
+		eCode := uint8(BGP_ERROR_UPDATE_MESSAGE_ERROR)
+		eSubCode := uint8(BGP_ERROR_SUB_INVALID_NEXT_HOP_ATTRIBUTE)
+		return NewMessageError(eCode, eSubCode, nil, "invalid nexthop address")
+	}
 	return nil
 }
 

--- a/pkg/packet/bgp/bgp_test.go
+++ b/pkg/packet/bgp/bgp_test.go
@@ -1582,6 +1582,120 @@ func Test_PathAttributeNextHop(t *testing.T) {
 	f("2001:db8::68")
 }
 
+func Test_PathAttributeNextHop_InvalidLength(t *testing.T) {
+	tests := []struct {
+		name       string
+		data       []byte
+		wantErr    bool
+		errSubCode uint8
+	}{
+		{
+			name: "valid IPv4",
+			data: []byte{
+				0x40, 0x03, // Flags, Type (NEXT_HOP)
+				0x04,         // Length = 4
+				192, 0, 2, 1, // 192.0.2.1
+			},
+			wantErr: false,
+		},
+		{
+			name: "valid IPv6",
+			data: []byte{
+				0x40, 0x03, // Flags, Type (NEXT_HOP)
+				0x10, // Length = 16
+				0x20, 0x01, 0x0d, 0xb8, 0, 0, 0, 0,
+				0, 0, 0, 0, 0, 0, 0x00, 0x68, // 2001:db8::68
+			},
+			wantErr: false,
+		},
+		{
+			name: "length 0",
+			data: []byte{
+				0x40, 0x03, // Flags, Type
+				0x00, // Length = 0
+			},
+			wantErr:    true,
+			errSubCode: BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR,
+		},
+		{
+			name: "length 1",
+			data: []byte{
+				0x40, 0x03, // Flags, Type
+				0x01, // Length = 1
+				192,  // partial data
+			},
+			wantErr:    true,
+			errSubCode: BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR,
+		},
+		{
+			name: "length 2",
+			data: []byte{
+				0x40, 0x03, // Flags, Type
+				0x02,   // Length = 2
+				192, 0, // partial data
+			},
+			wantErr:    true,
+			errSubCode: BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR,
+		},
+		{
+			name: "length 3",
+			data: []byte{
+				0x40, 0x03, // Flags, Type
+				0x03,      // Length = 3
+				192, 0, 2, // partial data
+			},
+			wantErr:    true,
+			errSubCode: BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR,
+		},
+		{
+			name: "length 5 (invalid)",
+			data: []byte{
+				0x40, 0x03, // Flags, Type
+				0x05,            // Length = 5
+				192, 0, 2, 1, 0, // 5 bytes
+			},
+			wantErr:    true,
+			errSubCode: BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR,
+		},
+		{
+			name: "length mismatch (header says 4, data is 2)",
+			data: []byte{
+				0x40, 0x03, // Flags, Type
+				0x04,   // Length = 4
+				192, 0, // only 2 bytes
+			},
+			wantErr:    true,
+			errSubCode: BGP_ERROR_SUB_ATTRIBUTE_LENGTH_ERROR,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			p := &PathAttributeNextHop{}
+			err := p.DecodeFromBytes(tt.data)
+
+			if tt.wantErr {
+				if err == nil {
+					t.Errorf("expected error but got none")
+					return
+				}
+				// Check error type
+				if msgErr, ok := err.(*MessageError); ok {
+					if msgErr.SubTypeCode != tt.errSubCode {
+						t.Errorf("expected error subcode %d, got %d", tt.errSubCode, msgErr.SubTypeCode)
+					}
+				} else {
+					t.Errorf("expected MessageError, got %T", err)
+				}
+			} else {
+				if err != nil {
+					t.Errorf("unexpected error: %v", err)
+				}
+			}
+		})
+	}
+}
+
 func Test_LsTLVDecode(t *testing.T) {
 	assert := assert.New(t)
 

